certora/ArrangerConduit.spec

```diff
@@ -0,0 +1,710 @@
+    assert isBrokerOtherAssetAfter == isBrokerOtherAssetBefore, "setBroker did not keep unchanged the rest of brokers for other assets";
+}
+
+// Verify revert rules on setBroker
+rule setBroker_revert(address usr, address asset, bool valid) {
+    env e;
+
+    mathint wardsSender = wards(e.msg.sender);
+
+    setBroker@withrevert(e, usr, asset, valid);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = wardsSender != 1;
+
+    assert revert1 => lastReverted, "revert1 failed";
+    assert revert2 => lastReverted, "revert2 failed";
+    assert lastReverted => revert1 || revert2, "Revert rules are not covering all the cases";
+}
+
+// Verify correct storage changes for non reverting deposit
+rule deposit(bytes32 ilk, address asset, uint256 amount) {
+    env e;
+
+    require asset == gem;
+
+    address buffer = registry.buffers(ilk);
+    require buffer != currentContract;
+
+    mathint balanceOfBufferBefore = gem.balanceOf(buffer);
+    mathint balanceOfConduitBefore = gem.balanceOf(currentContract);
+    mathint depositsBefore = deposits(asset, ilk);
+    mathint totalDepositsBefore = totalDeposits(asset);
+
+    require balanceOfBufferBefore + balanceOfConduitBefore <= max_uint256;
+
+    deposit(e, ilk, asset, amount);
+
+    mathint balanceOfBufferAfter = gem.balanceOf(buffer);
+    mathint balanceOfConduitAfter = gem.balanceOf(currentContract);
+    mathint depositsAfter = deposits(asset, ilk);
+    mathint totalDepositsAfter = totalDeposits(asset);
+
+    assert balanceOfBufferAfter == balanceOfBufferBefore - amount, "deposit did not decrease balance of buffer by amount";
+    assert balanceOfConduitAfter == balanceOfConduitBefore + amount, "deposit did not increase balance of conduit by amount";
+    assert depositsAfter == depositsBefore + amount, "deposit did not increase deposits by amount";
+    assert totalDepositsAfter == totalDepositsBefore + amount, "deposit did not increase totalDeposits by amount";
+}
+
+// Verify revert rules on deposit
+rule deposit_revert(bytes32 ilk, address asset, uint256 amount) {
+    env e;
+
+    require asset == gem;
+
+    address buffer = registry.buffers(ilk);
+    require buffer != currentContract;
+
+    bool canCall = roles.canCall(ilk, e.msg.sender, currentContract, to_bytes4(0xd954863c)); // deposit(bytes32,address,uint256)
+    mathint balanceOfBuffer = gem.balanceOf(buffer);
+    mathint allowanceBuffer = gem.allowance(buffer, currentContract);
+    mathint deposits = deposits(asset, ilk);
+    mathint totalDeposits= totalDeposits(asset);
+
+    deposit@withrevert(e, ilk, asset, amount);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = !canCall;
+    bool revert3 = balanceOfBuffer < to_mathint(amount);
+    bool revert4 = allowanceBuffer < to_mathint(amount);
+    bool revert5 = deposits + amount > max_uint256;
+    bool revert6 = totalDeposits + amount > max_uint256;
+    bool revert7 = buffer == 0;
+
+    assert lastReverted <=> revert1 || revert2 || revert3 ||
```

Comment:
Couldn't actually run it yet due to the submodule issue, but weird we are not having a revert condition about overflow on gem balance for the conduit address.

Reply:
Yeah that seems to be because the erc20 mock cannot overflow during transfers:

<img width="737" alt="image" src="https://github.com/makerdao/dss-conduits/assets/130504305/ff20b8ca-425e-48cd-a155-783c0225231d">


I tested with  the `unchecked` removed from `_transfer()` and can confirm that the extra revert condition you mention is indeed required in that case.

Reply:
Ok yeah 👍